wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/jpa/JpaWorkspaceDao.java

```diff
@@ -262,6 +262,19 @@ protected WorkspaceImpl doUpdate(WorkspaceImpl update) throws NotFoundException
     return merged;
   }
 
+  @Override
+  @Transactional
```

Comment:
Does a single select need to be transactional?

Reply:
all methods, including reading marked as transactional.
I thing the answer -  yes. They are.

Reply:
My follow up question would then be - why? :)

Reply:
I'm not sure but I feel that you're probably right.  The alternative to transactional annotation would be  - usage of `com.google.inject.persist.UnitOfWork`. Which we are using in the test s but not in code. As a result, the code will like this.
```java
  @Inject private UnitOfWork uow;
....
  @Override
  public long getTotalCount() throws ServerException {
    uow.begin();
    try {
      return managerProvider
          .get()
          .createNamedQuery("Workspace.getTotalCount", Long.class)
          .getSingleResult();
    } catch (RuntimeException x) {
      throw new ServerException(x.getLocalizedMessage(), x);
    } finally {
      uow.end();
    }
  }
```
I'm just not sure that even we waste some resources with Transactional it's a good idea to change an approach.  It can break something bigger than just a metric. Do you think it make sense to start digging in `UnitOfWork && JpaYYYDao` direction?

Reply:
I think what @metlos is trying to say, that single `SELECT Count ...` is already atomic operation and does not need transaction handling on any level, because it's ensured by database itself. So basically just remove the `@Transactional` annotation from your current code. Am I right @metlos ?

Reply:
>So basically just remove the @Transactional annotation from your current code

 - that is not going to work. You'll get on next JPA operation
```
java.lang.IllegalStateException: Work already begun on this thread. Looks like you have called UnitOfWork.begin() twice without a balancing call to end() in between.

	at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
	at com.google.inject.persist.jpa.JpaPersistService.begin(JpaPersistService.java:73)
	at org.eclipse.che.commons.test.tck.repository.JpaTckRepository.removeAll(JpaTckRepository.java:76)
	at org.eclipse.che.api.workspace.server.spi.tck.WorkspaceDaoTest.removeEntities(WorkspaceDaoTest.java:105)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:133)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:62)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:340)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:294)
	at org.testng.internal.TestInvoker.runConfigMethods(TestInvoker.java:683)
	at org.testng.internal.TestInvoker.runAfterGroupsConfigurations(TestInvoker.java:658)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:642)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:172)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:804)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:145)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)

```

Reply:
Ok, so let's succumb to Guice, I was just wondering why it is there. Btw, does that happen only in tests or also at runtime?

Reply:
this can potentially happen in both tests at runtime